### PR TITLE
48 rename entries to notes

### LINKED ISSUE #48: Rename entries to notes
Discussion: There are viz-app specific endpoints for entries, nodes, etc. I'd appreciate a simplification and would like to remove those endpoints. Does the discover app rely on those endpoints?

### PR DESCRIPTION
**Pull request by cspindler**
- [x] rename tables, models and endpoints
- [x] make location of notes optional
- [x] add public flag
- [x] change field name 'name' to 'title'
- [x] add field for KeyCloak user_sub
- [ ] update documentation

### Requires changes to:

- [x] deploy.mitwelten.org (tags list)
- [ ] walk.mitwelten.org (forms, etc.)
- [ ] discover.mitwelten.org

### Testing:

on the database server:

- clone the repo onto the database server and check out this branch:
    ```bash
    git clone https://github.com/mitwelten/mitwelten-db-backend.git
    cd mitwelten-db-backend
    git checkout origin/48-rename-entries-to-notes
    ```
- create `credentials.py` with `mitwelten_admin` as database user
- run the desired migration procedure (see [NOTES.md](https://github.com/mitwelten/mitwelten-db-backend/blob/d23c4efada5cd8ea8c192fa905f9b28f5e1a9272/schema/NOTES.md#migration-v23-to-v24))

on your machine:

- clone the repo onto your machine and check out this branch (see above)
- create `credentials.py` with `mitwelten_internal` as database user and `dev` as schema
- activate and install venv:
    ```bash
    cd services/api
    python -m venv .venv
    source .venv/bin/activate
    pip install -r requirements
    ```
- run the REST API service: 
    ```bash
    cd services
    ./api/.venv/bin/uvicorn api.main:app --reload
    ```

### Discussions:

There are viz-app specific endpoints for entries, nodes, etc. I'd appreciate a simplification and would like to remove those endpoints. Does the discover app rely on those endpoints? PR #50 implements this simplification.

### CODE REVIEW

**Comment by cspindler** [comment]
For testing: notes, tags and crossrefs can be copied to `dev` schema using `schema/migrations/migrate_v2_3__v2_4_dev.py`. Don't forget to truncate the tables before:

```sql
truncate dev.mm_tags_notes restart identity cascade;
truncate dev.tags restart identity cascade;
truncate dev.notes restart identity cascade;
truncate dev.files_note restart identity cascade;
```

---

Changes to the entries router are visible in all commits to `entries.py` except the file-renaming commit 9dfb1d18ec513bbf187a4d23c3ca9facf9f7359a

**Review by cspindler — COMMENTED** [review]
Check if PR #50 can be implemented to remove/consolidate the remaining viz endpoints

File: services/api/routers/tags.py
Comment (cspindler) [inline services/api/routers/tags.py:1]: Check if PR https://github.com/mitwelten/mitwelten-db-backend/pull/50 can be implemented to remove/consolidate the remaining viz endpoints

File: services/api/routers/notes.py
Comment (cspindler) [inline services/api/routers/notes.py:1]: - Add user_id to user_sub field on PUT(insert only)
- Resolve user_id to full name on GET

File: services/api/routers/notes.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+from fastapi import APIRouter, Depends, Query, HTTPException
+from sqlalchemy.sql import and_, between, select, text, func
+
+router = APIRouter(tags=['notes', 'discover'])
+
+# ------------------------------------------------------------------------------
+# NOTES
+# ------------------------------------------------------------------------------
+
+@router.get('/notes', response_model=List[Note], response_model_exclude_none=True)
```
Comment (cspindler) [inline services/api/routers/notes.py]: Check if authenticated, and filter according to `public` attribute

File: services/api/routers/tags.py
Comment (cspindler) [reply services/api/routers/tags.py:1]: Was implemented

File: services/api/routers/notes.py
Comment (cspindler) [reply services/api/routers/notes.py:1]: Also public flag  is added to the model and can now be set on insert / patch. Only users with 'internal' role can create public notes, others always create non-public notes.

**Review by cspindler — COMMENTED** [review]
Checked through, found all requirements to be met

File: services/api/routers/notes.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+    result = await database.fetch_all(query=query)
+
+    if result == None or len(result) == 0:
+        raise HTTPException(status_code=404, detail='Note not found')
+    else:
+        f_l = unique_everseen(result, lambda x: x['file_id'])
+        t_l = unique_everseen(result, lambda x: x['tag_id'])
+        e = dict(result[0])
+        e['files'] = [{'name': f['file_name'], 'link': f['object_name'], 'type': f['file_type']} for f in f_l if f['file_id'] != None]
```
Comment (andriwild) [inline services/api/routers/notes.py]: Change `'link'` to `'object_name'`

File: services/api/routers/notes.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    try:
+        check = await database.fetch_one(notes.select().where(notes.c.note_id == note_id))
+        if check == None:
+            raise HTTPException(status_code=404, detail='Note not found')
+
+        existing_by_id = None
+        existing_by_name = None
+        insert_tag = None
+
+        if body.id:
```
Comment (andriwild) [inline services/api/routers/notes.py]: Change `body.id` to `body.tag_id`

File: services/api/routers/notes.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    '''
+    delete_id = None
+    if body.name:
+        existing = await database.fetch_one(tags.select().where(tags.c.name == body.name))
+        if existing == None:
+            return
+        else:
+            delete_id = existing.tag_id
+
+    if body.id:
```
Comment (andriwild) [inline services/api/routers/notes.py]: Change `body.id` to `body.tag_id`  (also on line 259)

File: services/api/routers/notes.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        await transaction.rollback()
+        raise e
+    else:
+        await transaction.commit()
+
+
+@router.delete('/note/{note_id}/tag', dependencies=[Depends(AuthenticationChecker())], response_model=None, tags=['tags'])
+async def delete_tag_from_note(note_id: int, body: Tag) -> None:
+    '''
+    Deletes a tag from an note
```
Comment (andriwild) [inline services/api/routers/notes.py]: Typo: an note → a note

File: services/api/routers/notes.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    query = notes.update().returning(notes, notes.c.created_at.label('date')).where(notes.c.note_id == note_id).\
+        values({**update_data, notes.c.updated_at: func.current_timestamp()})
+
+    return await database.fetch_one(query)
+
+
+@router.delete('/note/{note_id}', response_model=None, dependencies=[Depends(AuthenticationChecker())])
+async def delete_note(note_id: int) -> None:
+    '''
+    ## Deletes an note
```
Comment (andriwild) [inline services/api/routers/notes.py]: Typo: an note → a note

File: services/api/routers/notes.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        t_l = unique_everseen(result, lambda x: x['tag_id'])
+        e = dict(result[0])
+        e['files'] = [{'name': f['file_name'], 'link': f['object_name'], 'type': f['file_type']} for f in f_l if f['file_id'] != None]
+        e['tags'] = [{'id': t['tag_id'], 'name': t['tag_name']} for t in t_l if t['tag_id'] != None]
+        return e
+
+@router.patch('/note/{note_id}', response_model=Note, dependencies=[Depends(AuthenticationChecker())])
+async def update_note(note_id: int, body: PatchNote = ..., auth = Depends(get_user)) -> Note:
+    '''
+    ## Updates an note
```
Comment (andriwild) [inline services/api/routers/notes.py]: Typo: an note → a note

File: services/api/routers/notes.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        await transaction.rollback()
+        raise e
+    else:
+        await transaction.commit()
+
+
+@router.post('/note/{note_id}/tag',tags=['tags'], response_model=None, dependencies=[Depends(AuthenticationChecker())], responses={'404': {'model': ApiErrorResponse}})
+async def add_tag_to_note(note_id: int, body: Tag) -> None:
+    '''
+    Adds a tag for an note
```
Comment (andriwild) [inline services/api/routers/notes.py]: Typo: an note → a note

File: services/api/routers/notes.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        delete_id = body.id
+
+    await database.execute(mm_tags_notes.delete().where(
+        and_(mm_tags_notes.c.tags_tag_id == delete_id, mm_tags_notes.c.notes_note_id == note_id)))
+
+
+@router.post('/note/{note_id}/file', dependencies=[Depends(AuthenticationChecker(['internal']))], response_model=None)
+async def add_file_to_note(note_id: int, body: File) -> None:
+    '''
+    Adds a file for an note
```
Comment (andriwild) [inline services/api/routers/notes.py]: Typo: an note → a note

File: services/api/routers/minio.py [hunk 10]
```diff
@@ -88,10 +102,27 @@ async def post_upload(file: UploadFile):
     except S3Error as e:
         if e.code != 'NoSuchKey':
             raise e
+    else:
+        return { 'object_name': stat.object_name, 'etag': stat.etag }
     # upload
     upload = storage.put_object(crd.minio.bucket, file.filename, file.file, length=-1, part_size=10*1024*1024)
     return { 'object_name': upload.object_name, 'etag': upload.etag }
 
+@router.post('/files/discover/', dependencies=[Depends(AuthenticationChecker(['internal']))])
```
Comment (andriwild) [inline services/api/routers/minio.py]: Remove `/` at the end of the path

**Review by andriwild — CHANGES_REQUESTED** [review]

**Pull request merged**